assert with message disables power assertion output

With a feature method like this

```
def one() {
	expect:
	("one".capitalize() * 3).toUpperCase() == "OneOneOne"
}
```

we get an error message like that:
```
Condition not satisfied:

("one".capitalize() * 3).toUpperCase() == "OneOneOne"
       |            |    |             |
       One          |    ONEONEONE     false
                    OneOneOne          6 differences (33% similarity)
                                       O(NE)O(NE)O(NE)
                                       O(ne)O(ne)O(ne)

Expected :OneOneOne

Actual   :ONEONEONE
```

The same is true if we explicitly use `assert` at the beginning of the condition. So far, so good.

Now this is what happens when using `assert` with an additional error message:

```
def one() {
	expect:
	assert ("one".capitalize() * 3).toUpperCase() == "OneOneOne", "My custom message"
}
```

The very helpful power assert output vanishes.
```
Condition not satisfied:

("one".capitalize() * 3).toUpperCase() == "OneOneOne"

My custom message
```

What I would like to see instead is the custom message prepended to the power assert output like this:

```
My custom message

Condition not satisfied:

("one".capitalize() * 3).toUpperCase() == "OneOneOne"
       |            |    |             |
       One          |    ONEONEONE     false
                    OneOneOne          6 differences (33% similarity)
                                       O(NE)O(NE)O(NE)
                                       O(ne)O(ne)O(ne)

Expected :OneOneOne

Actual   :ONEONEONE
```

Would it be possible to add a corresponding AST transformation for that kind of syntax?

I still see the need to be able to suppress a message, especially if we are using a third-party library, that provides its own exception messages.

```java
 resultActions.andExpect(status().isOk())
                .andExpect(view().name('login'))
                .andExpect(content().string(
                CoreMatchers.allOf(
                        containsString('<input type="hidden" name="service" value="foo" />'),
                        containsString('<input type="hidden" name="successURL" value="https:/login/success?kmsi=false" />'),
                        containsString('<input type="hidden" name="loginFailedURL" value="https:/login?failed=true&amp;login_hint=username@web.de" />'),
                        containsString('<input type="hidden" name="loginErrorURL" value="https:/login/error" />'),
                        containsString('<input type="hidden" name="statistics" value="foo" />'),
                        containsString('<input type="hidden" name="username" value="username@web.de" />'),
                        matcher
                )))
```

renders as

```
Condition failed with Exception:

resultActions.andExpect(status().isOk()) .andExpect(view().name('login')) .andExpect(content().string( CoreMatchers.allOf( containsString('<input type="hidden" name="service" value="foo" />'), containsString('<input type="hidden" name="successURL" value="https:/login/success?kmsi=false" />'), containsString('<input type="hidden" name="loginFailedURL" value="https:/login?failed=true&amp;login_hint=username@web.de" />'), containsString('<input type="hidden" name="loginErrorURL" value="https:/login/error" />'), containsString('<input type="hidden" name="statistics" value="foo" />'), containsString('<input type="hidden" name="username" value="username@web.de" />'), matcher )))
|             |         |        |        |         |      |
|             |         |        |        |         |      <org.springframework.test.web.servlet.result.ViewResultMatchers$$Lambda$382/219130824@77f4038c arg$1=login>
|             |         |        |        |         <org.springframework.test.web.servlet.result.ViewResultMatchers@27a90ce5>
|             |         |        |        java.lang.AssertionError: No ModelAndView found
|             |         |        <org.springframework.test.web.servlet.result.StatusResultMatchers$$Lambda$381/1314050802@51869ad6 arg$1=200>
|             |         <org.springframework.test.web.servlet.result.StatusResultMatchers@36511772>
|             <org.springframework.test.web.servlet.MockMvc$1@795faad val$mvcResult=org.springframework.test.web.servlet.DefaultMvcResult@1eb9d69a this$0=org.springframework.test.web.servlet.MockMvc@73385d3f>
<org.springframework.test.web.servlet.MockMvc$1@795faad val$mvcResult=org.springframework.test.web.servlet.DefaultMvcResult@1eb9d69a this$0=org.springframework.test.web.servlet.MockMvc@73385d3f>
```
The spring exception message is listed as cause.

with `assert <expectation> : 'MVC Expectation Failed'`  we get
```
Condition failed with Exception:

resultActions.andExpect(status().isOk()) .andExpect(view().name('login')) .andExpect(content().string( CoreMatchers.allOf( containsString('<input type="hidden" name="service" value="foo" />'), containsString('<input type="hidden" name="successURL" value="https:/login/success?kmsi=false" />'), containsString('<input type="hidden" name="loginFailedURL" value="https:/login?failed=true&amp;login_hint=username@web.de" />'), containsString('<input type="hidden" name="loginErrorURL" value="https:/login/error" />'), containsString('<input type="hidden" name="statistics" value="foo" />'), containsString('<input type="hidden" name="username" value="username@web.de" />'), matcher )))

MVC Expectation Failed
```

I would argue it would be event better to hide the complete expression. So if we change `assert exp : message` then we need something for this, any ideas?

Well, that's why my last sentence was:

> Please consider implementing it and making it the default behaviour - _**with an option to deactivate it**_ if you think it is necessary.

So providing an option to users to override the default behaviour (power-assert output) is, as I said, completely fine with me - not that you need my approval. ;-)

The sample log output you just showed might be a case in which a user would want to avoid power assert output, even though I think that it is rather exotic to use an extra libraries for assertions even though Spock provides its nice condition auto-assert which is no less expressive and easy to use. BTW, I would still like to have the power assert output if I could have the additional text I want. Why? The first line is identical anyway, but looks way uglier than it would using Spock conditions. Now we cannot even clearly see which of the many assertions failed. The power assert output at least helps a bit here, indicating that `.andExpect(view().name('login'))` failed with `java.lang.AssertionError: No ModelAndView found`. So actually it is a case for wanting power assert, not for avoiding it. It helps me more than the general information "MVC expectation failed". Of course it did, the power assert output tells me already.

As for how to implement the feature, I wrote something about it in a previous comment, too:

> I am sure there must be a way to make it configurable to **not** have power assert output if someone does not want it. An annotation-driven extension, `SpockConfig.groovy` or a combination of both, default config plus override per spec or feature method. Maybe even an override per (block of) condition(s) via `with(SomeCategory)` would be possible, but I am not the implementation expert here.

The idea that was mentioned above about different behaviour for `String` vs. `GString` would be quite counter-intuitive, I think we should not go that way. Another similar option might be a syntax like `assert condition : ["Additional message"]` which could be used like `assert condition : ["First paragraph", "Second paragraph"]`. I.e. if there was some kind of collection or array as an argument, Spock could use power assert output. It feels a bit ugly and definitely would need explicit documentation, but so would all other options.

I am just thinking aloud here and am really open to other ways of implementing or using this. Even if it was just a workaround like `powerAssert condition , "message"` it would be better than now because at least I could use it. I still would be happier with making power assert the default and providing ways to disable it, see my quote above.

--- a/docs/release_notes.adoc
+++ b/docs/release_notes.adoc
@@ -4,6 +4,36 @@ include::include.adoc[]
 
 == 1.3 (tbd)
 
+Potential breaking change: assertions with explicit messages now include power assertions output.
+
+.Given
+[source,groovy]
+----
+def a = 1
+def b = 2
+assert a == b : "Additional message"
+----
+
+.Before
+----
+a == b
+
+Additional message
+----
+
+.Now
+----
+a == b
+| |  |
+1 |  2
+  false
+
+Additional message
+----
+
+If you relied on this behavior to hide some output, or to prevent a stack overflow due to a self referenceing
+data structure, then you need to move the condition into a separate method that just returns the boolean result.
+
 === What's New In This release
 
 == 1.2 (2018-09-23)

--- a/spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java
+++ b/spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java
@@ -580,14 +580,9 @@ public class ConditionRewriter extends AbstractExpressionConverter<Expression> {
   private Statement rewriteMethodCondition(MethodCallExpression condition, Expression message, boolean explicit) {
     MethodCallExpression rewritten;
     int lastVariableNum;
-    if (message == null){
-      final Expression converted = convert(condition);
-      rewritten = (MethodCallExpression) unrecord(converted);
-      lastVariableNum = extractVariableNumber(converted);
-    }else{
-      rewritten = condition;
-      lastVariableNum = -1;
-    }
+    final Expression converted = convert(condition);
+    rewritten = (MethodCallExpression) unrecord(converted);
+    lastVariableNum = extractVariableNumber(converted);
 
     List<Expression> args = new ArrayList<>();
     args.add(rewritten.getObjectExpression());
@@ -612,14 +607,9 @@ public class ConditionRewriter extends AbstractExpressionConverter<Expression> {
       boolean explicit) {
     StaticMethodCallExpression rewritten;
     int lastVariableNum;
-    if (message == null){
-      final Expression converted = convert(condition);
-      rewritten = (StaticMethodCallExpression) unrecord(converted);
-      lastVariableNum = extractVariableNumber(converted);
-    }else{
-      rewritten = condition;
-      lastVariableNum = -1;
-    }
+    final Expression converted = convert(condition);
+    rewritten = (StaticMethodCallExpression) unrecord(converted);
+    lastVariableNum = extractVariableNumber(converted);
 
     List<Expression> args = new ArrayList<>();
     args.add(new ClassExpression(rewritten.getOwnerType()));
@@ -641,7 +631,7 @@ public class ConditionRewriter extends AbstractExpressionConverter<Expression> {
   }
 
   private Statement rewriteOtherCondition(Expression condition, Expression message) {
-    Expression rewritten = message == null ? convert(condition) : condition;
+    Expression rewritten = convert(condition);
 
     final Expression executeAndVerify = rewriteToSpockRuntimeCall(resources.getAstNodeCache().SpockRuntime_VerifyCondition,
         condition, message, Collections.singletonList(rewritten));
@@ -664,7 +654,7 @@ public class ConditionRewriter extends AbstractExpressionConverter<Expression> {
                     resources.getAstNodeCache().SpockRuntime_ConditionFailedWithException,
                     new ArgumentListExpression(Arrays.asList(
                         new VariableExpression(SpockNames.ERROR_COLLECTOR),
-                        message == null ? new VariableExpression(SpockNames.VALUE_RECORDER) : ConstantExpression.NULL, // recorder
+                        new VariableExpression(SpockNames.VALUE_RECORDER), // recorder
                         new ConstantExpression(resources.getSourceText(condition)),                                 // text
                         new ConstantExpression(condition.getLineNumber()),                                          // line
                         new ConstantExpression(condition.getColumnNumber()),                                        // column
@@ -711,12 +701,10 @@ public class ConditionRewriter extends AbstractExpressionConverter<Expression> {
         new ArgumentListExpression(args));
 
     args.add(new VariableExpression(SpockNames.ERROR_COLLECTOR, resources.getAstNodeCache().ErrorCollector));
-    args.add(message == null ?
-        AstUtil.createDirectMethodCall(
+    args.add(AstUtil.createDirectMethodCall(
             new VariableExpression(SpockNames.VALUE_RECORDER),
             resources.getAstNodeCache().ValueRecorder_Reset,
-            ArgumentListExpression.EMPTY_ARGUMENTS) :
-        ConstantExpression.NULL);
+            ArgumentListExpression.EMPTY_ARGUMENTS));
     args.add(new ConstantExpression(resources.getSourceText(condition)));
     args.add(new ConstantExpression(condition.getLineNumber()));
     args.add(new ConstantExpression(condition.getColumnNumber()));

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/condition/ConditionRendering.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/condition/ConditionRendering.groovy
@@ -762,12 +762,12 @@ BLOCKED == 0
   // for implicit closure calls see ImplicitClosureCallRendering
   void "explicit closure call"() {
     def func = { it }
-
+    expect:
     isRendered """
 func.call(42) == null
 |    |        |
 |    42       false
-${func.toString()}
+${func.dump()}
     """, {
       assert func.call(42) == null
     }

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/condition/ExplicitConditionsWithMessage.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/condition/ExplicitConditionsWithMessage.groovy
@@ -24,7 +24,7 @@ import spock.lang.Issue
 /**
  * @author Peter Niederwieser
  */
-@Issue("http://issues.spockframework.org/detail?id=22")
+@Issue(["https://github.com/spockframework/spock/issues/145", "https://github.com/spockframework/spock/issues/922"])
 class ExplicitConditionsWithMessage extends ConditionRenderingSpec {
   def "evaluation of satisfied condition"() {
     expect:
@@ -41,6 +41,8 @@ class ExplicitConditionsWithMessage extends ConditionRenderingSpec {
     expect:
     isRendered """
 1 + 2 == 2
+  |   |
+  3   false
 
 need to brush up my math
     """, {
@@ -52,6 +54,8 @@ need to brush up my math
     expect:
     isRendered """
 a + b == 2
+| | | |
+1 3 2 false
 
 a: 1 b: 2
     """, {
@@ -65,6 +69,10 @@ a: 1 b: 2
     expect:
     isRendered """
 map.a + map.b == 2
+|   | | |   | |
+|   1 3 |   2 false
+|       [a:1, b:2]
+[a:1, b:2]
 
 [a:1, b:2]
     """, {
@@ -90,9 +98,60 @@ map.a + map.b == 2
     expect:
     isRendered """
 1 + 2 == 2
+  |   |
+  3   false
     """, {
       def x = null
       assert 1 + 2 == 2, x
     }
   }
+
+  void "explicit assertion with message"() {
+    expect:
+    isRendered """
+a == b
+| |  |
+1 |  2
+  false
+
+Additional message
+    """, {
+      def a = 1
+      def b = 2
+      assert a == b : "Additional message"
+    }
+  }
+
+  void "explicit assertion with methodCondition and message"() {
+    expect:
+    isRendered """
+a.contains(b)
+| |        |
+| false    bar
+foo
+
+Additional message
+    """, {
+      def a = 'foo'
+      def b = 'bar'
+      assert a.contains(b) : "Additional message"
+    }
+  }
+
+
+  void "explicit assertion with static methodCondition and message"() {
+    expect:
+    isRendered """
+Boolean.parseBoolean(a)
+|       |            |
+|       false        foo
+class java.lang.Boolean
+
+Additional message
+    """, {
+      def a = 'foo'
+      assert Boolean.parseBoolean(a) : "Additional message"
+    }
+  }
+
 }

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/condition/MatcherConditionRendering.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/condition/MatcherConditionRendering.groovy
@@ -68,6 +68,9 @@ Expected: <43>
 
   @IsRendered("""
 that(x, equalTo(43))
+|    |
+|    42
+false
 
 oops!
 

